spring-data-mongodb/src/main/java/org/springframework/data/mongodb/core/ReactiveMongoTemplate.java

```diff
@@ -0,0 +1,2449 @@
+	}
+
+	protected <T> Flux<T> doInsertAll(Collection<? extends T> listToSave, MongoWriter<Object> writer) {
+
+		final Map<String, List<T>> elementsByCollection = new HashMap<String, List<T>>();
+
+		listToSave.forEach(element -> {
+			MongoPersistentEntity<?> entity = mappingContext.getPersistentEntity(element.getClass());
+
+			if (entity == null) {
+				throw new InvalidDataAccessApiUsageException("No PersistentEntity information found for " + element.getClass());
+			}
+
+			String collection = entity.getCollection();
+			List<T> collectionElements = elementsByCollection.get(collection);
+
+			if (null == collectionElements) {
+				collectionElements = new ArrayList<T>();
+				elementsByCollection.put(collection, collectionElements);
+			}
+
+			collectionElements.add(element);
+		});
+
+		return Flux.fromIterable(elementsByCollection.keySet())
+				.flatMap(collectionName -> doInsertBatch(collectionName, elementsByCollection.get(collectionName), writer));
+	}
+
+	protected <T> Flux<T> doInsertBatch(final String collectionName, final Collection<? extends T> batchToSave,
+										final MongoWriter<Object> writer) {
+
+		Assert.notNull(writer);
+
+		Mono<List<Tuple2<T, Document>>> prepareDocuments = Flux.fromIterable(batchToSave)
+				.flatMap(new Function<T, Flux<Tuple2<T, Document>>>() {
+					@Override
+					public Flux<Tuple2<T, Document>> apply(T o) {
+
+						initializeVersionProperty(o);
+						maybeEmitEvent(new BeforeConvertEvent<T>(o, collectionName));
+
+						Document dbDoc = toDbObject(o, writer);
+
+						maybeEmitEvent(new BeforeSaveEvent<T>(o, dbDoc, collectionName));
+						return Flux.zip(Mono.just(o), Mono.just(dbDoc));
+					}
+				}).collectList();
+
+		Flux<Tuple2<T, Document>> insertDocuments = prepareDocuments.flatMap(tuples -> {
+
+			List<Document> dbObjects = tuples.stream().map(Tuple2::getT2).collect(Collectors.toList());
+
+			return insertDocumentList(collectionName, dbObjects).thenMany(Flux.fromIterable(tuples));
+		});
+
+		return insertDocuments.map(tuple -> {
+
+			populateIdIfNecessary(tuple.getT1(), tuple.getT2().get(ID_FIELD));
+			maybeEmitEvent(new AfterSaveEvent<T>(tuple.getT1(), tuple.getT2(), collectionName));
+			return tuple.getT1();
+		});
+	}
+
+	/* (non-Javadoc)
+	 * @see org.springframework.data.mongodb.core.ReactiveMongoOperations#save(reactor.core.publisher.Mono)
+	 */
+	@Override
+	public <T> Mono<T> save(Mono<? extends T> objectToSave) {
+		return objectToSave.then(this::save);
+	}
+
+	/* (non-Javadoc)
+	 * @see org.springframework.data.mongodb.core.ReactiveMongoOperations#save(reactor.core.publisher.Mono, java.lang.String)
+	 */
+	@Override
+	public <T> Mono<T> save(Mono<? extends T> objectToSave, String collectionName) {
+		return objectToSave.then(o -> save(o, collectionName));
+	}
+
+	/* (non-Javadoc)
+	 * @see org.springframework.data.mongodb.core.ReactiveMongoOperations#save(java.lang.Object)
+	 */
+	public <T> Mono<T> save(T objectToSave) {
```

Comment:
This one should check that `T` is not a collection/`Publisher`. Otherwise, it's very easy to pass in objects that are not intended for saving (`save(Flux.just(…)`).